dao/src/main/java/se/fortnox/reactivewizard/db/ReactiveStatementFactory.java

```diff
@@ -92,46 +90,54 @@ private Mono<Object> getResultMono(StatementContext statementContext) {
         }));
     }
 
-    /**
-     * Create Flux statement.
-     *
-     * @param args                the arguments
-     * @param connectionScheduler the scheduler
-     * @return the Flux statement
-     */
-    public Object create(Object[] args, ConnectionScheduler connectionScheduler) {
-        StatementContext statementContext = new StatementContext(() -> statementFactory.create(args), connectionScheduler);
-        if (isReturnTypeMono) {
-            Mono<Object> resultMono = getResultMono(statementContext);
-            if (shouldAddDebugErrorHandling()) {
-                Exception queryFailure = new RuntimeException(QUERY_FAILED);
-                resultMono = resultMono.onErrorResume(thrown -> {
-                    queryFailure.initCause(thrown);
-                    return Mono.error(queryFailure);
-                });
-            }
-            resultMono = Mono.from(metrics.measure(resultMono, this::logSlowQuery));
-            return decorated(resultConverter.apply(resultMono), statementContext);
-        } else {
-            Flux<Object> resultFlux = getResultFlux(statementContext);
-            if (shouldAddDebugErrorHandling()) {
-                Exception queryFailure = new RuntimeException(QUERY_FAILED);
-                resultFlux = resultFlux.onErrorResume(thrown -> {
-                    queryFailure.initCause(thrown);
-                    return Flux.error(queryFailure);
-                });
-            }
-            resultFlux = pagingOutput.apply(resultFlux, args);
-            resultFlux = Flux.from(metrics.measure(resultFlux, this::logSlowQuery));
-            resultFlux = resultFlux.onBackpressureBuffer(RECORD_BUFFER_SIZE);
-            return decorated(resultConverter.apply(resultFlux), statementContext);
+    public <T> Mono<T> createMono(
+        Metrics metrics,
+        Supplier<Statement> statementSupplier
+    ) {
+        var statementContext = new StatementContext(statementSupplier, connectionScheduler);
+        Mono<T> resultMono = getResultMono(statementContext);
+        if (shouldAddDebugErrorHandling()) {
+            resultMono = resultMono.onErrorResume(thrown ->
+                Mono.error(new RuntimeException(QUERY_FAILED, thrown))
```

Comment:
> Will this not result in creating new RuntimeExceptions for each error, as opposed as the old code reusing the same one?

In the old code, each time method `ReactiveStatementFactory#create` is called, the new instance of exception is created.
The only difference is that now exception is created only if error happens while previously it was created even if error didn't happened.

The reusability of the exception instance could only happen in the old code if lambda provided to method `onErrorResume` is called several times. I suppose this may happen if we add a retry. Is there any benefit in reusing the exception instance in this case?